plugin/rewrite/reverter.go

```diff
@@ -21,22 +22,77 @@ type ResponseRule struct {
 // dig will complain with ';; Question section mismatch: got example.org/HINFO/IN'
 type ResponseReverter struct {
 	dns.ResponseWriter
+	request          *dns.Msg
 	originalQuestion dns.Question
+	originalOPT      *dns.OPT
 	ResponseRewrite  bool
 	ResponseRules    []ResponseRule
 }
 
+// return a deep copy of the OPT record
+func dupOPT(opt *dns.OPT) (o *dns.OPT) {
+	edns0 := make([]dns.EDNS0, len(opt.Option))
+	for i, e := range opt.Option {
+		edns0[i] = e.Copy()
+	}
+	o = &dns.OPT{opt.Hdr, edns0}
+	return o
+}
+
+// return a deep copy of the OPT inside this msg
+func copyOPT(m *dns.Msg) (o *dns.OPT) {
```

Comment:
this func is not needed. You can use existing `IsEdns0()` method and then make a copy if needed

Reply:
No. OPT has no public Copy().
However, now, EDNS0 has not neither.
so the last way to have a copy of OPT is : msg.Copy().isEdns0().

Reply:
I meant one can use code like
```
o := dupOPT(msg.IsEnds0())
```
Though, `dupOPT()` should expect nil parameter for this to work. Otherwise, a bit more cumbersome construction will have to be used
```
o := msg.IsEnds0()
if o != nil {
  o := dupOPT(o)
}
```